File: simulator/spells/scorching_ray.py
from simulator.actions.action_types import BonusAction
from simulator.battle_map import Map, map_position_toggled_cache
from simulator.spells.spell import SpellStats
from simulator.misc import DamageType, avg_roll, Conditions, Visibility
from simulator.actions.actoid import Actoid, FactoryFlags, ActoidFlags
from functools import cache
from simulator.threat_utils import mean_dmg
from simulator.threat_interfaces import DirectThreat, DirectThreatFactory
from itertools import combinations_with_replacement
import logging
from simulator.utils.roll_types import RollType, ROLL_TYPE_DELTA, ROLL_TYPE_CRIT_DELTA, ThreatModifierType

logger = logging.getLogger("EncounTroll")

class ScorchingRayFactory(DirectThreatFactory):
    level = 2
    range = SpellStats.Range.FEET_120.value
    target = SpellStats.Target.THREE_CREATURES
    duration = SpellStats.Duration.INSTANTANEOUS
    concentration = False
    type = SpellStats.Type.HARMFUL
    dc = None
    dmg_type = DamageType.Fire

    def __init__(self, to_hit, action_type, caster):
        super().__init__()
        self.flags |= FactoryFlags.IS_ATTACK_LIKE
        self.to_hit = to_hit
        self.action_type = action_type  # SCORCHING_RAY, QUICKENED_SCORCHING_RAY
        self.dmg_dice = '2d6'
        self.combatant = caster

    def __str__(self):
        """
        Important for FSM building
        """
        return "ScorchingRayFactory"

    def get_quickened_kwargs(self):
        return {'to_hit': self.to_hit, 'caster': self.combatant}

    def get_eligible_targets(self):
        swallower = self.combatant.get_swallower()
        if swallower:
            return [swallower, swallower, swallower]
        # Range is so big that it doesn't matter
        return combinations_with_replacement([e for e in Map.get().get_enemies(self.combatant) if not e.is_affected_by(Conditions.SWALLOWED)], 3)

    def create_all(self):
        targets = self.get_eligible_targets()
        return [ScorchingRay(t, self) for t in targets]

    def create(self, targets):
        return ScorchingRay(targets, self)

    def calculate_threat_to_target(self, target, **kwargs):
        battle_map = Map.get()
        if battle_map.get_cartesian_distance(self.combatant, target) <= ScorchingRayFactory.range:
            roll_type = RollType.STRAIGHT if not battle_map.is_enemy_adjacent(self.combatant) else RollType.DISADVANTAGE
            to_hit_total = self.to_hit + ROLL_TYPE_DELTA[roll_type][max(0, min(target.ac - self.to_hit, 20))]
            return 3 * mean_dmg(to_hit_total, self.dmg_dice, 0, target.ac, ROLL_TYPE_CRIT_DELTA[roll_type], target.is_resistant_to(ScorchingRayFactory.dmg_type))
        else:
            return 0

    def calculate_threat_to_target_delta_single_target(self, target, modifiers):
        """
        Helper function
        """
        mod_to_hit_flat = modifiers.get(ThreatModifierType.TO_HIT_FLAT, 0)
        mod_to_hit_die = modifiers.get(ThreatModifierType.TO_HIT_DIE, '0d0')
        roll_type = modifiers.get(ThreatModifierType.ROLL_TYPE, RollType.STRAIGHT)

        to_hit_total = self.to_hit + mod_to_hit_flat + avg_roll(mod_to_hit_die)
        to_hit_total += ROLL_TYPE_DELTA[roll_type][max(0, min(target.ac - to_hit_total, 20))]
        total_crit = ROLL_TYPE_CRIT_DELTA[roll_type]

        # We assume the maximum threat in case where all three rays are aimed at the target
        return 3*(mean_dmg(to_hit_total, self.dmg_dice, 0, target.ac, total_crit, target.is_resistant_to(ScorchingRayFactory.dmg_type)) - \
            mean_dmg(self.to_hit, self.dmg_dice, 0, target.ac, 1, target.is_resistant_to(ScorchingRayFactory.dmg_type)))

    def calculate_threat_to_target_delta(self, target, modifiers, *args, **kwargs):
        """
        Calculates the threat delta of the factory to a specific target given stat modifications.
        This is useful calculating the potential reduction of threat_in caused by abilities of enemies, e.g. advantage on saving throw
        against fireball or bane on attack rolls etc.
        """
        # We assume the maximum threat in case where all three rays are aimed at the target
        return 3 * self.calculate_threat_to_target_delta_single_target(target, modifiers)

    def calculate_max_threat(self):
        swallower = self.combatant.get_swallower()
        if swallower:
            targets = [swallower]
        else:
            targets = [e for e in Map.get().get_enemies(self.combatant) if not e.is_affected_by(Conditions.SWALLOWED)]
        return max([self.calculate_threat_to_target(t) for t in targets])


class ScorchingRay(Actoid, DirectThreat):

    def __init__(self, targets, factory, **kwargs):
        Actoid.__init__(self, actoid_flags=ActoidFlags.IS_SPELL | ActoidFlags.IS_ATTACK_LIKE | ActoidFlags.IS_DIRECT_THREAT)
        self.targets = targets
        self.factory = factory
        self.empowered = kwargs.get("empowered", False)
        self.roll_type = RollType.STRAIGHT

    def __str__(self):
        return ("Quickened " if self.factory.action_type is BonusAction.QUICKENED_SCORCHING_RAY else "") + f"Scorching Ray on {self.targets[0]}, {self.targets[1]} and {self.targets[2]}"

    def shorthand_str(self):
        return ("Quickened " if self.factory.action_type is BonusAction.QUICKENED_SCORCHING_RAY else "") + "Scorching Ray"

    @map_position_toggled_cache
    def calculate_threat(self, **kwargs):
        battle_map = Map.get()
        roll_type = RollType.STRAIGHT if not battle_map.is_enemy_adjacent(self.factory.combatant) else RollType.DISADVANTAGE
        crit_multiplier = ROLL_TYPE_CRIT_DELTA[roll_type]
        to_hit_total = self.factory.to_hit + ROLL_TYPE_DELTA[roll_type][max(0, min(self.targets[0].ac - self.factory.to_hit, 20))]
        dmg_acc = mean_dmg(to_hit_total, self.factory.dmg_dice, 0, self.targets[0].ac, crit_multiplier, self.targets[0].is_resistant_to(ScorchingRayFactory.dmg_type))
        to_hit_total = self.factory.to_hit + ROLL_TYPE_DELTA[roll_type][max(0, min(self.targets[1].ac - self.factory.to_hit, 20))]
        dmg_acc += mean_dmg(to_hit_total, self.factory.dmg_dice, 0, self.targets[1].ac, crit_multiplier, self.targets[1].is_resistant_to(ScorchingRayFactory.dmg_type))
        to_hit_total = self.factory.to_hit + ROLL_TYPE_DELTA[roll_type][max(0, min(self.targets[2].ac - self.factory.to_hit, 20))]
        dmg_acc += mean_dmg(to_hit_total, self.factory.dmg_dice, 0, self.targets[2].ac, crit_multiplier, self.targets[2].is_resistant_to(ScorchingRayFactory.dmg_type))
        return dmg_acc

    def clear_cache(self):
        self.calculate_threat.cache_clear()

    def calculate_threat_delta(self, modifiers, *args, **kwargs):
        ret = self.factory.calculate_threat_to_target_delta_single_target(self.targets[0], modifiers)
        ret += self.factory.calculate_threat_to_target_delta_single_target(self.targets[1], modifiers)
        ret += self.factory.calculate_threat_to_target_delta_single_target(self.targets[2], modifiers)
        return ret

    def get_eligible_coords(self, distances, shortest_paths):
        battle_map = Map.get()
        coords_for_first = battle_map.get_free_coords_in_cartesian_range(battle_map.get_combatant_position(self.targets[0]),
                                                                        distances,
                                                                        inflate_to_size=self.factory.combatant.size,
                                                                        rng=ScorchingRayFactory.range,
                                                                        combatant=self.factory.combatant)
        coords_for_second = battle_map.get_free_coords_in_cartesian_range(battle_map.get_combatant_position(self.targets[1]),
                                                                          distances,
                                                                          inflate_to_size=self.factory.combatant.size,
                                                                          rng=ScorchingRayFactory.range,
                                                                          combatant=self.factory.combatant)
        coords_for_third = battle_map.get_free_coords_in_cartesian_range(battle_map.get_combatant_position(self.targets[2]),
                                                                          distances,
                                                                          inflate_to_size=self.factory.combatant.size,
                                                                          rng=ScorchingRayFactory.range,
                                                                          combatant=self.factory.combatant)
        free_coords_in_range = coords_for_third.intersection(coords_for_first.intersection(coords_for_second))

        return {coord for coord in free_coords_in_range if
                battle_map.visibility_dict_for_all_coords[coord][self.targets[0]] is not Visibility.NONE
                and battle_map.visibility_dict_for_all_coords[coord][self.targets[1]] is not Visibility.NONE
                and battle_map.visibility_dict_for_all_coords[coord][self.targets[2]] is not Visibility.NONE}

    def is_current_coord_eligible(self):
        if all([t is self.factory.combatant.get_swallower() for t in self.targets]):
            return True
        battle_map = Map.get()
        return battle_map.get_cartesian_distance(self.factory.combatant, self.targets[0]) <= ScorchingRayFactory.range \
            and battle_map.get_cartesian_distance(self.factory.combatant, self.targets[1]) <= ScorchingRayFactory.range \
            and battle_map.get_cartesian_distance(self.factory.combatant, self.targets[2]) <= ScorchingRayFactory.range

File: c++/spells/scorching_ray.cpp
#include "spells/scorching_ray.hpp"
#include "core/combatant.hpp"
#include "core/battle_map.hpp"
#include <memory>
#include <limits>

namespace enc
{

  ScorchingRayFactory::ScorchingRayFactory(int toHit, AbilityType abilityType, Combatant *caster, Resource *resource)
      : DirectThreatFactory("ScorchingRayFactory", "Scorching Ray", caster, abilityType), _toHit(toHit), _resource(resource),
        _dmgDice(ScorchingRayFactory::rayDmgDice)
  {
    setFlag(FactoryFlags::IS_ATTACK_LIKE);
  }

  std::vector<Combatant *> ScorchingRayFactory::getEligibleTargets() const
  {
    BattleMap &battleMap = BattleMap::getInstance();
    Combatant *swallower = _combatant->getSwallower();
    if(swallower)
      {
        return {swallower};
      }
    return battleMap.getNonSwallowedEnemiesWithinRadius(_combatant, static_cast<int>(ScorchingRayFactory::range));
  }

  std::vector<std::shared_ptr<Actoid>> ScorchingRayFactory::createAll(void *previousActionInDag)
  {
    auto eligibleTargets = getEligibleTargets();
    std::vector<std::shared_ptr<Actoid>> result;
    result.reserve(eligibleTargets.size());
    for(const auto &target : eligibleTargets)
      {
        result.push_back(std::make_shared<ScorchingRay>(*target, *this));
      }
    return result;
  }

  std::shared_ptr<Actoid> ScorchingRayFactory::create(void *target)
  {
    return std::make_shared<ScorchingRay>(*static_cast<Combatant *>(target), *this);
  }

  double ScorchingRayFactory::calculateThreatToTarget(Combatant *target, const Kwargs &kwargs) const
  {
    BattleMap &battleMap = BattleMap::getInstance();
    if(target->getSwallower())
      {
        return 0;
      }
    if(battleMap.getCartesianDistanceCombatants(*_combatant, *target) <= static_cast<int>(ScorchingRayFactory::range))
      {
        auto rollType = battleMap.isEnemyAdjacent(*_combatant) ? RollType::DISADVANTAGE : RollType::STRAIGHT;
        int acDifference = std::max(0, std::min(20, target->getAC() - _toHit));
        int toHitTotal = _toHit + ROLL_TYPE_DELTA.at(rollType).at(acDifference);
        double singleRay = meanDmg(toHitTotal, {_dmgDice}, 0, target->getAC(), target->isImmuneTo(ScorchingRayFactory::dmgType),
                                   target->isResistantTo(ScorchingRayFactory::dmgType), ROLL_TYPE_CRIT_DELTA.at(rollType));
        return ScorchingRayFactory::numRays * singleRay;
      }

    return 0;
  }

  double ScorchingRayFactory::calculateThreatToTargetDelta(Combatant *target, const ThreatModifiers &modifiers) const
  {
    if(target->isImmuneTo(ScorchingRayFactory::dmgType))
      {
        return 0;
      }

    int modToHitFlat = modifiers.getOrDefault(ThreatModifierType::TO_HIT_FLAT, 0);
    Die modToHitDie = modifiers.getOrDefault(ThreatModifierType::TO_HIT_DIE, Die{0, 0});
    RollType rollType = modifiers.getOrDefault(ThreatModifierType::ROLL_TYPE, RollType::STRAIGHT);
    int targetAC = modifiers.getOrDefault(ThreatModifierType::TARGET_AC, 0);

    int totalTargetAC = targetAC + target->getAC();
    double toHitTotal = _toHit + modToHitFlat + avgRoll(modToHitDie);
    int needToRollAtLeast = std::max(0, std::min(totalTargetAC - static_cast<int>(toHitTotal), 20));
    toHitTotal += ROLL_TYPE_DELTA.at(rollType).at(needToRollAtLeast);
    double totalCrit = ROLL_TYPE_CRIT_DELTA.at(rollType);

    double modifiedThreat = ScorchingRayFactory::numRays
                            * meanDmg(toHitTotal, {_dmgDice}, 0, totalTargetAC, target->isImmuneTo(ScorchingRayFactory::dmgType),
                                      target->isResistantTo(ScorchingRayFactory::dmgType), totalCrit);

    double originalThreat = ScorchingRayFactory::numRays
                            * meanDmg(_toHit, {_dmgDice}, 0, target->getAC(), target->isImmuneTo(ScorchingRayFactory::dmgType),
                                      target->isResistantTo(ScorchingRayFactory::dmgType), 1);

    return modifiedThreat - originalThreat;
  }

  double ScorchingRayFactory::calculateMaxThreat() const
  {
    auto eligibleTargets = getEligibleTargets();
    double maxThreat = std::numeric_limits<double>::lowest();
    for(const auto &target : eligibleTargets)
      {
        double threat = ScorchingRay(*target, *this).calculateThreat(Kwargs());
        maxThreat = std::max(maxThreat, threat);
      }
    return eligibleTargets.empty() ? 0 : maxThreat;
  }

  std::string ScorchingRay::toString() const
  {
    std::string prefix = (_factory._abilityType == AbilityType::QUICKENED_SCORCHING_RAY) ? "Quickened " : "";
    return prefix + "Scorching Ray at " + _target._name;
  }

  std::string ScorchingRay::shorthandStr() const
  {
    std::string prefix = (_factory._abilityType == AbilityType::QUICKENED_SCORCHING_RAY) ? "Quickened " : "";
    return prefix + "Scorching Ray";
  }

  double ScorchingRay::calculateThreat(const Kwargs &kwargs)
  {
    BattleMap &battleMap = BattleMap::getInstance();
    auto rollType = battleMap.isEnemyAdjacent(*_factory._combatant) ? RollType::DISADVANTAGE : RollType::STRAIGHT;
    int acDifference = std::max(0, std::min(20, _target.getAC() - _factory._toHit));
    int toHitTotal = _factory._toHit + ROLL_TYPE_DELTA.at(rollType).at(acDifference);
    double singleRay = meanDmg(toHitTotal, {_factory._dmgDice}, 0, _target.getAC(), _target.isImmuneTo(ScorchingRayFactory::dmgType),
                               _target.isResistantTo(ScorchingRayFactory::dmgType), ROLL_TYPE_CRIT_DELTA.at(rollType));
    return ScorchingRayFactory::numRays * singleRay;
  }

  double ScorchingRay::calculateThreatDelta(const ThreatModifiers &modifiers) const
  {
    return _factory.calculateThreatToTargetDelta(&_target, modifiers);
  }

  std::optional<CoordVector>
  ScorchingRay::getEligibleCoords(const blaze::DynamicVector<int> &distances, const blaze::DynamicMatrix<Coord> &shortestPaths)
  {
    ScorchingRayFactory &factory = dynamic_cast<ScorchingRayFactory &>(getFactory());
    BattleMap &battleMap = BattleMap::getInstance();
    Combatant *swallower = factory._combatant->getSwallower();
    Coord currCoord = battleMap.getCombatantCoordinates(*factory._combatant).getRoot();
    if(swallower)
      {
        if(swallower == &_target)
          {
            CoordVector coords = {currCoord};
            return coords;
          }
        return {};
      }

    if(!factory._combatant->isAffectedByAny({Conditions::GRAPPLED, Conditions::GRAPPLING, Conditions::RESTRAINED}))
      {
        return battleMap.getFreeCoordsInCartesianRange(battleMap.getCombatantCoordinates(_target).get(), distances, factory._combatant->getSize(),
                                                       static_cast<int>(ScorchingRayFactory::range), factory._combatant->_instanceId);
      }
    else if(battleMap.getCartesianDistanceCombatants(*factory._combatant, _target) <= static_cast<int>(ScorchingRayFactory::range))
      {
        CoordVector coords = {currCoord};
        return coords;
      }
    return {};
  }
}
